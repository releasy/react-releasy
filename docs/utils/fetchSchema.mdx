---
name: fetchSchema
menu: Utils
route: /utils/fetchSchema
---

# fetchSchema

## Usage

```javascript
import { fetchSchema } from 'react-releasy';

fetchSchema({
  url: 'https://yourserveraddress.com/graphql',
  dir: './data',
}).then(() => {
  console.log('Schema downloaded');
  process.exit();
});

```
