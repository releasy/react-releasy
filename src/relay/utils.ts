import { CacheConfig, RequestNode } from 'relay-runtime';

export const isMutation = (request: RequestNode) => request.operationKind === 'mutation';

export const mustForceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);
