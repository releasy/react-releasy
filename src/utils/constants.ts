export const IS_NODE = typeof window === 'undefined';

export const WINDOW_DATA_KEY = '__RELEASY_DATA__';
