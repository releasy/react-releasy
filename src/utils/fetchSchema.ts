import {
  buildClientSchema,
  printSchema,
  introspectionQuery,
} from 'graphql';
import fse from 'fs-extra';
import fetch from 'node-fetch';

type FetchSchemaInput = {
  url: string,
  path: string,
};

const fetchSchema = async (input: FetchSchemaInput): Promise<void> => {
  const { url, path } = input;

  const response = await fetch(url, {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query: introspectionQuery,
    }),
  });

  const result = await response.json();

  const clientSchema = buildClientSchema(result.data);
  const schema = printSchema(clientSchema);

  fse.ensureFile(path);

  return fse.writeFile(path, schema);
};

export default fetchSchema;
