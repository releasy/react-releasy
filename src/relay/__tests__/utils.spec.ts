import {
  createCacheConfig,
  createMutationRequest,
  createQueryRequest,
} from '../../../tests/fixtures/relay';

import { isMutation, mustForceFetch } from '../utils';

it('should verify if is a mutation', async () => {
  const request = createMutationRequest();

  expect(isMutation(request)).toMatchSnapshot();
});

it('should verify if is not a mutation', async () => {
  const request = createQueryRequest();

  expect(isMutation(request)).toMatchSnapshot();
});

it('should verify if must force a fetch', async () => {
  const cacheConfig = createCacheConfig({ force: true });

  expect(mustForceFetch(cacheConfig)).toMatchSnapshot();
});

it('should verify if must not force a fetch', async () => {
  const cacheConfig = createCacheConfig({ force: false });

  expect(mustForceFetch(cacheConfig)).toMatchSnapshot();
});
