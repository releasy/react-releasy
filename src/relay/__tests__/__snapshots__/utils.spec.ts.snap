// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should verify if is a mutation 1`] = `true`;

exports[`should verify if is not a mutation 1`] = `false`;

exports[`should verify if must force a fetch 1`] = `true`;

exports[`should verify if must not force a fetch 1`] = `false`;
